plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1167 @@
+        __int32         timeoutWait = 100;  //!< Amount of time to wait between retries
+        __int32         attemptNum = 0;
+
+
+        if (shouldRead && (maxRecords <= 0 || consumedRecCount < maxRecords))
+        {
+            RdKafka::Message*   messageObjPtr = NULL;
+
+            while (!messageObjPtr && shouldRead && attemptNum < maxAttempts)
+            {
+                messageObjPtr = consumerPtr->getOneMessage(); // messageObjPtr must be deleted when we are through with it
+
+                if (messageObjPtr)
+                {
+                    try
+                    {
+                        switch (messageObjPtr->err())
+                        {
+                            case RdKafka::ERR_NO_ERROR:
+                                {
+                                    RtlDynamicRowBuilder    rowBuilder(resultAllocator);
+                                    unsigned                len = sizeof(__int32) + sizeof(__int64) + sizeof(size32_t) + messageObjPtr->len();
+                                    byte*                   row = rowBuilder.ensureCapacity(len, NULL);
+
+                                    // Populating this structure:
+                                    //  EXPORT KafkaMessage := RECORD
+                                    //      UNSIGNED4   partitionNum;
+                                    //      UNSIGNED8   offset;
+                                    //      STRING      message;
+                                    //  END;
+
+                                    *(__int32*)(row) = messageObjPtr->partition();
+                                    *(__int64*)(row + sizeof(__int32)) = messageObjPtr->offset();
+                                    *(size32_t*)(row + sizeof(__int32) + sizeof(__int64)) = messageObjPtr->len();
+                                    memcpy(row + sizeof(__int32) + sizeof(__int64) + sizeof(size32_t), messageObjPtr->payload(), messageObjPtr->len());
+
+                                    result = rowBuilder.finalizeRowClear(len);
+
+                                    lastMsgOffset = messageObjPtr->offset();
+                                    ++consumedRecCount;
+
+                                    // Give opportunity for consumer to pull in any additional messages
+                                    consumerPtr->handle()->poll(0);
+                                }
+                                break;
+
+                            case RdKafka::ERR__TIMED_OUT:
+                                // No new messages arrived and we timed out waiting
+                                ++attemptNum;
+                                consumerPtr->handle()->poll(timeoutWait);
+                                break;
+
+                            case RdKafka::ERR__PARTITION_EOF:
+                                // We reached the end of the messages in the partition
+                                #ifdef _DEBUG
```

Comment:
I've added an initial version of a log context-based traceLevel logging.  What is passed around is the traceLevel initially set, rather than the context, as some of the objects live long past the initial call.  For the same reason, I've left DBGLOG() in place rather than attempting to log through a logging context object.  Please let me know if my assumptions regarding logging context lifetime are wrong or if there is a better way to handle these.